### PR TITLE
Add notifications aa1 to master

### PR DESCRIPTION
**Pull request by Aichnerc**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently there is no way to get notifications for chat messages.
- Currently there is a bug that caused empty chats to be on top of all chats.
- Currently there is a bug that prohibits people from going offline in our presence system.
- Currently the message in the text field of a chat is being deleted when the component unmounts.

**What is the new behavior (if this is a feature change)?**
- Now users can receive notifications for chat messages.
- Now empty chats are being sorted by their creation date.
- Users now go offline when signing out.
- The message is now stored in localStorage and loaded when re-opening the chat.

**Other information**:
- The version has been updated to `v3.1.0`.

### CODE REVIEW

File: src/components/molecules/MessageItem/index.jsx [hunk 1]
```diff
@@ -122,9 +151,72 @@ class MessageItem extends React.Component {
     }
   };
 
+  processMsg = (msg) => {
+    let replacedText, replacePattern1, replacePattern2, replacePattern3;
+    let inputText = msg.escape();
```
Comment (Theralun) [inline src/components/molecules/MessageItem/index.jsx:156]: This could be a `const`.

File: src/components/molecules/MessageItem/index.jsx [hunk 2]
```diff
@@ -122,9 +151,72 @@ class MessageItem extends React.Component {
...
+    );
+
+    return replacedText;
+  };
+
+  checkForLink = (msg) => {
+    // eslint-disable-next-line no-useless-escape
+    let replacePattern1 = /(\b(https?|ftp):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gim;
+    // eslint-disable-next-line no-useless-escape
+    let replacePattern2 = /(^|[^\/])(www\.[\S]+(\b|$))/gim;
```
Comment (Theralun) [inline src/components/molecules/MessageItem/index.jsx:189]: These two could be `const`s as well

File: src/store/actions/notificationActions.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
+// Get array of all notifications for the user
+export const getNotifs = () => {
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const notifRef = firebase.database().ref("/notifications/");
+
+    const uid = firebase.auth().currentUser.uid;
```
Comment (Theralun) [inline src/store/actions/notificationActions.js:7]:
```suggestion
    // Retreive the currently logged in users uid
    const uid = firebase.auth().currentUser.uid;
```

File: src/store/actions/notificationActions.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    // Create reference
+    chatMessageRef.off("child_added");
+  };
+};
+
+// Create a new notification for each recipient
+export const createNotification = (details, recipients, chatName) => {
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const uid = firebase.auth().currentUser.uid;
```
Comment (Theralun) [inline src/store/actions/notificationActions.js:53]:
```suggestion

    // Retreive the currently logged in users uid
    const uid = firebase.auth().currentUser.uid;
```

File: src/store/actions/notificationActions.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    notifRef.remove().then(() => dispatch({ type: "DELETENOTIF_SUCCESS" }));
+  };
+};
+
+// Throw notification out of the DB by chat id
+export const removeNotifications = (chid) => {
+  return (dispatch, getState, { getFirebase, getFirestore }) => {
+    const firebase = getFirebase();
+    const uid = firebase.auth().currentUser.uid;
+    const notifRef = firebase.database().ref("/notifications/" + chid + uid);
```
Comment (Theralun) [inline src/store/actions/notificationActions.js:92]:
```suggestion
    const firebase = getFirebase();

    // Retreive the currently logged in users uid
    const uid = firebase.auth().currentUser.uid;

    const notifRef = firebase.database().ref("/notifications/" + chid + uid);
```

File: src/store/reducers/rootReducer.js [hunk 6]
```diff
@@ -28,6 +29,7 @@ const rootReducer = combineReducers({
   donMsg: donMsgReducer, // Donation messages
   comment: commentReducer, // Comments
   chat: chatReducer, // Chat
+  notifications: notificationReducer,
```
Comment (schettn) [inline src/store/reducers/rootReducer.js:32]: What about a comment?

File: src/store/actions/notificationActions.js (hunk 5 above)
Comment (schettn) [reply src/store/actions/notificationActions.js]: Why?

File: src/store/actions/notificationActions.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Update state to false
+    notifRef.remove().then(() => dispatch({ type: "DELETENOTIF_SUCCESS" }));
+  };
+};
+
+/**
+ * SPDX-License-Identifier: (EUPL-1.2)
+ * Copyright © 2019-2020 Werbeagentur Christian Aichner
+ */
```
Comment (schettn) [inline src/store/actions/notificationActions.js]: Is it common for you to not separate the exports in a section?

File: src/App.js [hunk 8]
```diff
@@ -126,12 +132,14 @@ class App extends React.Component {
...
+    notifications: state.notifications.notifications,
   };
 };
 
 const mapDispatchToProps = (dispatch) => {
   return {
     initPresenceHandler: (uid) => dispatch(initPresenceHandler(uid)),
+    getNotifs: (uid) => dispatch(getNotifs(uid)),
   };
 };
```
Comment (schettn) [inline src/App.js:144]: Looks like the "regioning" is missing in that file xD

File: src/components/molecules/MessageItem/index.jsx [hunk 9]
```diff
@@ -122,9 +151,72 @@ class MessageItem extends React.Component {
...
+      replacePattern3,
+      '<a href="mailto:$1">$1</a>'
+    );
+
+    return replacedText;
+  };
+
+  checkForLink = (msg) => {
+    // eslint-disable-next-line no-useless-escape
+    let replacePattern1 = /(\b(https?|ftp):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gim;
```
Comment (schettn) [inline src/components/molecules/MessageItem/index.jsx]: ReGeX iS sO eAsY.. @ReinerRabensteiner

File: src/store/actions/notificationActions.js (hunk 7 above)
Comment (Aichnerc) [reply src/store/actions/notificationActions.js]: Sections? Nope. That's what different actions are for.

File: src/store/reducers/rootReducer.js (hunk 6 above)
Comment (Aichnerc) [reply src/store/reducers/rootReducer.js:32]: Also consider to use inline comments.

```suggestion
  /* This is a comment */
  notifications: notificationReducer,
```

File: src/store/actions/notificationActions.js (hunk 7 above)
Comment (schettn) [reply src/store/actions/notificationActions.js]: sorry. Regions xD

```
export {
 foo,
 bar
}
```

Comment (Aichnerc) [reply src/store/actions/notificationActions.js]: No need for regions. Just exports in that.

**Pull request merged**